docs/rfc/rfc-002-zero-copy-encoding.md

```diff
@@ -0,0 +1,402 @@
+  runtime whether an enum value is unknown. An alternative would be to make adding new values to existing enums breaking
+* all enum values must be `<= 255`. Any enum in a blockchain application which needs more than 256 values is probably
+  doing something very wrong.
+* all oneof's must be the *only* element in their containing message and must start at field number `1` and be added in
+  continuous ascending order - this makes it possible to quickly check for unknown values
+* all `oneof` field numbers must be `<= 255`. Any `oneof` which needs more field cases is probably doing something very
+  wrong.
+
+These requirements make the encoding and generated code simpler.
+
+### Encoding
+
+#### Buffers and Memory Management
+
+By default, this encoding attempts to use a single fixed size encoding buffer of 64kb. This imposes a limit on the
+maximum size of a message that can be encoded. In the context of a message passing protocol for blockchains, this
+is generally a reasonable limit and the only known valid use case for exceeding it is to store user-uploaded byte
+code for execution in VMs. To accommodate this, large `string` and `bytes` values can be encoded in additional
+standalone buffers if needed. Still, the body of a message included all scalar and message fields
+must fit inside the 64kb buffer.
+
+While this design decision greatly simplifies the encoding and decoding logic, as well as the complexity of
+generated code, it does mean that APIs will need to do proper bounds checking when writing data that is not fixed
+size and return errors.
+
+The term `Root` is used to refer to the main 64kb buffer plus any additional large `string`/`bytes` buffers that are
+allocated.
+
+#### Scalar Encoding
+
+* `bool`s are encoded as 1 byte - `0` or `1`
+* `uint32`, `int32`, `sint32`, `fixed32`, `sfixed32` are encoded as 4 bytes by default
+* `uint64`, `int64`, `sint64`, `fixed64`, `sfixed64` are encoded as 8 bytes by default
+* `enum`s are encoded as 1 byte and values *MUST* be in the range of `0` to `255`.
+* all scalars declared as `optional` are prefixed with 1 additional byte whose value is `0` or `1` to indicate presence
+
+#### Message Encoding
+
+By default, messages field are encoded inline as structs. Meaning that if a message struct takes 8 bytes then its inline
+field in another struct will add 8 bytes to that struct size.
+
+`optional` message fields will be prefixed by 1 byte to indicate presence. (Alternatively, we could encode optional
+message fields as pointers (see below) if the desire is to save memory when they are rarely used needed.)
+
+#### Oneof’s
+
+`oneof`s are encoded as a combination of a `uint8` discriminant field and memory that is as large as the largest member
+field. `oneof` field numbers *MUST* be between `1` and `255`.
+
+```protobuf
+message Foo {
+  oneof sum {
+    bool x = 1;
+    int32 y = 2;
+  }
+}
+```
+
+A discriminant of `0` indicates that the field is not set.
```

Comment:
You're just asserting arbitrary rules about how protobuf types are encoded. Consumers of protobuf definitions are not required to follow these rules when encoding messages to the wire, and your code can't assume these rules have been followed when decoding protobuf messages from the wire.

Reply:
Why not? That's what a specification is for

Reply:
I think the RFC already makes this point pretty clear. To quote the first sentence:
> This RFC proposes a zero-copy encoding that is derived from the schema definitions defined in .proto files in the SDK
 and all app chains.

We are not calling it protobuf